lib/app/rendering.ts

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2025, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import type {Request, Response} from 'express';
+import _ from 'underscore';
+
+import {GoldenLayoutRootStruct} from '../clientstate-normalizer.js';
+import * as normalizer from '../clientstate-normalizer.js';
+import type {ShortLinkMetaData} from '../handlers/handler.interfaces.js';
+import * as utils from '../utils.js';
+import {
+    PugRequireHandler,
+    RenderConfig,
+    RenderConfigFunction,
+    ServerDependencies,
+    ServerOptions,
+} from './server.interfaces.js';
+import {isMobileViewer} from './url-handlers.js';
+
+/**
+ * Create rendering-related functions
+ * @param pugRequireHandler - Handler for Pug requires
+ * @param options - Server options
+ * @param dependencies - Server dependencies
+ * @returns Rendering functions
+ */
+export function createRenderHandlers(
+    pugRequireHandler: PugRequireHandler,
+    options: ServerOptions,
+    dependencies: ServerDependencies,
+): {
+    renderConfig: RenderConfigFunction;
+    renderGoldenLayout: (
+        config: GoldenLayoutRootStruct,
+        metadata: ShortLinkMetaData,
+        req: Request,
+        res: Response,
+    ) => void;
+    embeddedHandler: (req: Request, res: Response) => void;
+} {
```

Comment:
I would probably extract this object type into an interface

Reply:
We already had one 😮‍💨 - fixed.